pkg/agent/containerd/containerd.go

```diff
@@ -203,15 +214,74 @@ func preloadImages(ctx context.Context, cfg *config.Node) error {
 				continue
 			}
 		}
-		_, err = client.Import(ctxContainerD, imageReader, containerd.WithAllPlatforms(true))
+
+		images, err := client.Import(ctxContainerD, imageReader, containerd.WithAllPlatforms(true))
 		file.Close()
 		if err != nil {
 			logrus.Errorf("Unable to import %s: %v", filePath, err)
+			continue
+		}
+
+		err = retagImages(ctxContainerD, client, images, cfg.AgentConfig.AirgapExtraRegistry)
+		if err != nil {
+			logrus.Errorf("Unable to retag images imported from %s: %v", filePath, err)
 		}
 	}
 	return nil
 }
 
+// retagImages retags all listed images as having been pulled from the given remote registries.
+// If duplicate images exist, they are overwritten. This is most useful when using a private registry
+// for all images, as can be configured by the RKE2/Rancher system-default-registry setting.
+func retagImages(ctx context.Context, client *containerd.Client, images []images.Image, registries []string) error {
+	var errs []error
+	imageService := client.ImageService()
+	for _, image := range images {
+		name, err := parseNamedTagged(image.Name)
+		if err != nil {
+			errs = append(errs, errors.Wrap(err, "failed to parse image name"))
+			continue
+		}
+		logrus.Debugf("Imported %s", image.Name)
+		for _, registry := range registries {
+			image.Name = fmt.Sprintf("%s/%s:%s", registry, docker.Path(name), name.Tag())
+			if _, err = imageService.Create(ctx, image); err != nil {
```

Comment:
This looks fine to me but I think, to avoid super edge-case weirdness, you will want to leverage the lease facility: https://github.com/containerd/containerd/blob/master/docs/garbage-collection.md#how-to-use-leases